### PR TITLE
Inline embeds into Facia page

### PR DESCRIPTION
**Pull request by rich-nguyen**
By making the `facia-press` fetch well-formed embed json content, and `facia` inline embeds at render time, we can dramatically improve the visual performance of snaps. @janua @sammorrisdesign @mattosborn 
### Before

before
### After

after

### CODE REVIEW

**Comment by OliverJAsh** [comment]
This is so good.

On Wed, 20 Jan 2016, 18:19 Richard Nguyen <notifications@github.com> wrote:

> By making the facia-press fetch well-formed embed json content, and facia
> inline embeds at render time, we can dramatically improve the visual
> performance of snaps. @janua <https://github.com/janua> @sammorrisdesign
> <https://github.com/sammorrisdesign> @mattosborn
> <https://github.com/mattosborn>
> Before
>
> [image: before]
> <https://cloud.githubusercontent.com/assets/4549425/12458553/467f7314-bfa2-11e5-9bae-e00356eac658.gif>
> After
>
> [image: after]
> <https://cloud.githubusercontent.com/assets/4549425/12458558/4aee0fb4-bfa2-11e5-8fbc-3320671e8cab.gif>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/guardian/frontend/pull/11635
> Commit Summary
>
>    - Remove snapUri usages to simplify frontend
>    - Add a getCurated for snap hydration
>    - Provide a reason why facia-press is returning 503
>    - Add EnrichedContent to facia-press to enable inlining of embeds
>    - Extend SnapStuff with pre-fetched embed html content
>    - Add a test and switch to check embeds are inline
>
> File Changes
>
>    - *M* common/app/conf/switches/FaciaSwitches.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-0> (9)
>    - *M* common/app/layout/Sublink.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-1> (19)
>    - *M* common/app/model/Formats.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-2> (9)
>    - *M* common/app/model/SupportedUrl.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-3> (6)
>    - *M* common/app/model/pressedContent.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-4> (25)
>    - *M*
>    common/app/views/fragments/items/facia_cards/contentCard.scala.html
>    <https://github.com/guardian/frontend/pull/11635/files#diff-5> (31)
>    - *A*
>    data/pressedPage/3c487e70db2c1bb401604f2cb257768584a57748107e2234fd4a780eb9bd69e8
>    <https://github.com/guardian/frontend/pull/11635/files#diff-6> (1)
>    - *M* facia-press/app/controllers/Application.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-7> (2)
>    - *M* facia-press/app/frontpress/FapiFrontPress.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-8> (41)
>    - *M* facia/test/FaciaControllerTest.scala
>    <https://github.com/guardian/frontend/pull/11635/files#diff-9> (14)
>
> Patch Links:
>
>    - https://github.com/guardian/frontend/pull/11635.patch
>    - https://github.com/guardian/frontend/pull/11635.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/11635>.
>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

**Comment by stephanfowler** [comment]
Very nice. Another step towards no page-jumping...

**Comment by paperboyo** [comment]
:gem:

File: common/app/views/fragments/items/facia_cards/contentCard.scala.html [hunk 1]
```diff
@@ -19,21 +18,25 @@
...
         data-discussion-url="@item.header.url.get(request)#comments"
     }
+    data-link-name="@item.analyticsPrefix | card-@{ index + 1 }"
+    data-item-visibility="@visibilityDataAttribute"
+    data-test-id="facia-card"
+    @item.id.map { id => data-id="@id" }
+    @item.snapStuff.map(_.dataAttributes)
+    @item.shortUrl.map { shortUrl => data-loyalty-short-url="@shortUrl" }>
 
-data-link-name="@item.analyticsPrefix | card-@{
```
Comment (janua) [inline common/app/views/fragments/items/facia_cards/contentCard.scala.html]: You don't need the index anymore?

File: facia-press/app/frontpress/FapiFrontPress.scala [hunk 2]
```diff
@@ -119,6 +131,38 @@ trait FapiFrontPress extends QueryDefaults with Logging with ExecutionContexts {
...
+                case _ => content
+              }
+            } recover {
+              case _ => content
+            }
+            Response.Async.Right(updatedContent)
+          }
+        newContent.getOrElse(Response.Right(content))
+      }
+    case _ => Response.Right(pressed)
```
Comment (janua) [inline facia-press/app/frontpress/FapiFrontPress.scala]: In this method, do we want to count failures perhaps? Or log at least?

Comment (janua) [inline facia-press/app/frontpress/FapiFrontPress.scala]: Also, when the request falls back to the original `PressedContent`, what do we get on the page?

File: common/app/views/fragments/items/facia_cards/contentCard.scala.html (hunk 1 above)
Comment (rich-nguyen) [inline common/app/views/fragments/items/facia_cards/contentCard.scala.html]: I moved it to https://github.com/guardian/frontend/pull/11635/files#diff-a2b2ea169f137c124dce5857de42402fR21

**Comment by janua** [comment]
This is brilliant! My dreams have come true! :clap: 

:+1:

File: facia-press/app/frontpress/FapiFrontPress.scala (hunk 2 above)
Comment (rich-nguyen) [inline facia-press/app/frontpress/FapiFrontPress.scala]: We should log it, yes. The fallback behaviour is today's behaviour, so when original `PressedContent` is used, client-side javascript will enhance the content card.

**Comment by janua** [comment]
Things I should mention;
 - Will the JS be too eager for current thrashers or snaps?
 - Are the interactives team aware of the new context they will be writing for? (In terms of eager JS)
 - Will we ever be able to remove the JS surrounding snaps?

**Comment by rich-nguyen** [comment]
Questions 1+2: I've emailed the visuals team (and authors of thrashers), it will be turned off in prod until they give us the ok. 
Question 3: It remains the fallback until we are in a position to use the inline html system exclusively. It is also the only implementation that deals with embed types: `fragment` and `document`. We can extend the inlinig system (or contract the embed system) if we push for a unified pressing.

Extra: We should totally add comment counts to `EnrichedContent`

**Comment by janua** [comment]
Great!
Definitely add comment counts! :clap:

**Pull request merged**